### DISCUSSION TITLE
Sonoff TH origin THR316 Temperature/Heat/Humidity

### DISCUSSION THREAD

**Discussion by AdamaFla** [arendst/Tasmota#16841] [Q&A]
Hello,

I have recently bought several Sonoff TH origin THR316 devices and successfully flashed them with Tasmota, I cannot however, find any documentation on what to set the GPIO config for this device in order to get the correct reading for the attached THS01 rj11 sensor for Temperature/Heat/Humidity. It does display a Temperature by default, but it appears wrong by quite a lot.

Any help or suggestions please?

**Comment by barbudor** [comment]
Read the page down to the end, GPIO is given here: https://templates.blakadder.com/sonoff_THR316.html
Exact config depends on your sensor but standard Sonoff should use SI7021

**Reply by sfromis** [reply]
If you see a "ESP32 Temperature" in the web UI (and in the display), that is a bad number. In newest dev Tasmota builds (and soon release builds), this sensor is off by default, unless you enable it, but already with older builds you can deactivate the driver using the command `SetSensor127 0`.

**Comment by AdamaFla** [comment]
Thank you both!

**Reply by sfromis** [reply]
A very recent change (about 10 minutes ago) is the auto configuration setups for the device now including selection of temperature/humidity sensor type.

**Comment by MrSuicideParrot** [comment]
@AdamaFla Have you been able to use the THS01 by configuring the sensor as a SI7021? I tried to use the THS01 as a SI7021 in esphome and I couldn't read its values successfully.

**Reply by sfromis** [reply]
In Tasmota, I got THS01 to work by configuring gpio 25 as SI7021. However not very reliably after some time, now I'm getting a lot of these messages after setting `Weblog 3` for details....
`Pin25 timeout waiting for high pulse`
And in the JSON payloads `"SI7021":{"Temperature":null,"Humidity":null,"DewPoint":null}`
Not sure if this indicates issues in the sensor or in the driver.

**Reply by MrSuicideParrot** [reply]
But have you ever managed to get a correct measurement from the sensor?

**Reply by sfromis** [reply]
Yes, I do still get lots of reasonable (the word "correct" is too high a bar to decide on) measurements, but also many intermittent failures with the indicated symptoms. My first weeks were without reliability issues. Thus, I know that it *can* work with Tasmota, and with only one THS01, I have no good guess about why I do no longer have it working reliably. And I do have lots of other sensors :wink: 

Do note that a key observation is that I **always** gets *something*, null or valid. If you do not get either, you may have a more basic setup issue. Of course, I do not know how ESPHome reacts in case of read failures.....

Also beware that Tasmota configuration for SI7021 may be entirely different from what would work with ESPHome, as SI7021 usually is an I2C sensor, which Sonoff hides behind a very different onewire protocol. I'd suggest that you look towards how the Sonoff TH10/TH16 devices are configured around the temperature sensor, and likely do something similar to that.

**Reply by MrSuicideParrot** [reply]
Thank you very much for your advice.
I need to explore a little bit more... , and  if I'm not able to solve this problem I'll try tasmota.

Esphome has two si7021 sensors, one that is i2c and another for sonoff that works with one-wire, similar to the th20 family. I was using the one wire version because is the one that supposedly works with sonoff sensors, but the sensor wasn't detected. So, I thought this new sensor, ths01, could have a different protocol than the previous version. 
I opened the sensor and the temperature sensor inside (the smd component) seems different when compared with the photos I found online of the sonoff si7021.

**Reply by sfromis** [reply]
My point was that it strictly does not matter what the sensor chip is, as there also is a small MCU to transform the I2C protocol to something over onewire. Since Tasmota uses the "SI7021" config (name is arguably a bit misleading, but historical reasons) for both the Sonoff TH10/TH16 boards and the newer THR316/THR320 boards, that suggests that Tasmota is expecting the same behavior on the onewire bus, basically seeing "SI7021" as a name for the protocol used, rather than what chip is inside the probe.

I'm not much of a fan of the onewire bus, as there can be a lot of issues getting things working if the timing is a bit off, and ESP32 vs ESP8266 has speed differences, leading to possible timing differences.

On this board, the black 16-pin chip is that MCU, while the white U1 is the sensor: 
![image](https://user-images.githubusercontent.com/47082390/204917693-9909227b-aeb9-4f92-aa9e-f20170abab6f.png)

My THS01 also has a differently-looking sensor, and still that pesky MCU (if only Sonoff originally used all 4 wires in the plug, they could have made it straight I2C, but alas). 
![image](https://user-images.githubusercontent.com/47082390/204917975-853ae632-2f91-4480-ada4-7b3721532b93.png)

**Reply by MrSuicideParrot** [reply]
I was able to put this sensor working on esphome, I forgot to turning on the GPIO27 that works as VCC for this sensors.
 I was complicating it a bit... ahahah

**Comment by patman-cp** [comment]
I was thinking of getting one of these.

Do you have any notes or directions on how to flash it?

I've flashed some Sonoff power outlets to Tasmota.

And is it working OK for you with proper temperature readings and such? I wasn't sure about some of the comments here.

**Reply by MantasGSXR** [reply]
@patman-cp did you end up getting one? I just got one and flashed it without too much problems en even though it "works", the readings are more "null" than an actual reading. There also doesn't seem to be a pattern except that it's null much longer than it ever is on a reading.

I was wondering if this is normal behaviour or if I need to look for a "solution".

**Reply by patman-cp** [reply]
> @patman-cp did you end up getting one? 

Nope.

**Comment by toughvj** [comment]
@AdamaFla do you have original firmware for THr316 that you could share? I've removed mine accidentally and can't make firmware restore :/

**Comment by AllenSlaghuis** [comment]
Did anybody get it to work. i bought the oraganal sensor from Sonoff and it is a tsh01. and cant get it to work.

**Comment by sag007** [comment]
Hey Allen,

This one didn't work for me either, I could not manage to get any readings from the temp sensor at all. I tried Tasmota with different various sensors selected, also tried ESPhome but this didn't work either, so I ended up using Tasmota on the Origin unit and an alternate DS18B20 sensor I purchased off AliExpress. I fitted an RJ9 connector to it so it would plug straight into the unit and boom, the readings came up straight away and so now I can use it as I wanted for the control of my hot water heater.

I'm thinking that the guys in the above posts are correct in that there is a chip in the moulded part of the temp sensor cable that is changing the data so the available sensor in Tasmota do not work. I didn't try cutting this off the cable but as cheap as you can get these sensors for, it might be worth giving it a go using the sensor without this in between.

Nathan

**Comment by AllenSlaghuis** [comment]
Doy you think I can use a dht ? As I want to use it in my bathroom to turn on a extractor that is humidity based controll?

________________________________

**Reply by rhannink** [reply]
Hello,

I needed to put:
```DhtDelay 500,40```
in the autoexec.bat to get reliable readings from the sensor.

Gr. Remco

**Comment by ChrisVonFrosta** [comment]
which firmware need to choose? with default tasmota it ends in boot loop. also with every variant i tested

**Reply by sfromis** [reply]
That should automatically select an appropriate binary, meaning that the details around the boot loop is more important. Note that ESP32 devices will very often not be able to enable Wifi until you disconnect the USB/serial adapter, and power using mains instead.

**Reply by ChrisVonFrosta** [reply]
I just tried it before I saw that you had replied – lightning fast as usual! With 230 V connected, the Wi-Fi manager came up. Thank you!
